crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs

```diff
@@ -1,12 +1,80 @@
 use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
 use ruff_macros::{derive_message_formats, ViolationMetadata};
-use ruff_python_ast::{name::QualifiedName, Arguments, Expr, ExprAttribute, ExprCall};
+use ruff_python_ast::{
+    name::QualifiedName, Arguments, Expr, ExprAttribute, ExprCall, ExprContext, ExprName,
+    StmtClassDef,
+};
 use ruff_python_semantic::analyze::typing;
 use ruff_python_semantic::Modules;
+use ruff_python_semantic::ScopeKind;
 use ruff_text_size::Ranged;
 
 use crate::checkers::ast::Checker;
 
+fn is_airflow_builtin_or_provider(segments: &[&str], module: &str, symbol_suffix: &str) -> bool {
+    match segments {
+        ["airflow", "providers", rest @ ..] => {
+            if let (Some(pos), Some(last_element)) =
+                (rest.iter().position(|&s| s == module), rest.last())
+            {
+                pos + 1 < rest.len() && last_element.ends_with(symbol_suffix)
+            } else {
+                false
+            }
+        }
+
+        ["airflow", rest @ ..] => {
+            if let (Some(start_element), Some(last_element)) = (rest.first(), rest.last()) {
+                start_element.starts_with(module) & last_element.ends_with(symbol_suffix)
```

Comment:
Should the `starts_with` be instead a `start_element == module` as I think we need to match the module exactly?

Also, I think this should be `&&` and not `&` (bitwise AND) ;)

Reply:
The `&&` typo makes me wonder how come the tests failed to catch this

Reply:
I'm unable to push to this PR, so here's the change that I made locally to fix this and add some documentation:

```diff
diff --git a/crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs b/crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs
index 0f6831317..592329fde 100644
--- a/crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs
+++ b/crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs
@@ -11,21 +11,35 @@ use ruff_text_size::Ranged;
 
 use crate::checkers::ast::Checker;
 
+/// Check whether the segments corresponding to the fully qualified name points to a symbol that's
+/// either a builtin or coming from one of the providers in Airflow.
+///
+/// The pattern it looks for are:
+/// - `airflow.providers.**.<module>.**.*<symbol_suffix>` for providers
+/// - `airflow.<module>.**.*<symbol_suffix>` for builtins
+///
+/// where `**` is one or more segments separated by a dot, and `*` is one or more characters.
+///
+/// Examples for the above patterns:
+/// - `airflow.providers.google.cloud.secrets.secret_manager.CloudSecretManagerBackend` (provider)
+/// - `airflow.secrets.base_secrets.BaseSecretsBackend` (builtin)
 fn is_airflow_builtin_or_provider(segments: &[&str], module: &str, symbol_suffix: &str) -> bool {
     match segments {
         ["airflow", "providers", rest @ ..] => {
             if let (Some(pos), Some(last_element)) =
                 (rest.iter().position(|&s| s == module), rest.last())
             {
+                // Check that the module is not the last element i.e., there's a symbol that's
+                // being used from the `module` that ends with `symbol_suffix`.
                 pos + 1 < rest.len() && last_element.ends_with(symbol_suffix)
             } else {
                 false
             }
         }
 
-        ["airflow", rest @ ..] => {
-            if let (Some(start_element), Some(last_element)) = (rest.first(), rest.last()) {
-                start_element.starts_with(module) & last_element.ends_with(symbol_suffix)
+        ["airflow", first, rest @ ..] => {
+            if let Some(last) = rest.last() {
+                first == module && last.ends_with(symbol_suffix)
             } else {
                 false
             }
@@ -35,18 +49,26 @@ fn is_airflow_builtin_or_provider(segments: &[&str], module: &str, symbol_suffix
     }
 }
 
+/// Check whether the symbol is coming from the `secrets` builtin or provider module which ends
+/// with `Backend`.
 fn is_airflow_secret_backend(segments: &[&str]) -> bool {
     is_airflow_builtin_or_provider(segments, "secrets", "Backend")
 }
 
+/// Check whether the symbol is coming from the `hooks` builtin or provider module which ends
+/// with `Hook`.
 fn is_airflow_hook(segments: &[&str]) -> bool {
     is_airflow_builtin_or_provider(segments, "hooks", "Hook")
 }
 
+/// Check whether the symbol is coming from the `operators` builtin or provider module which ends
+/// with `Operator`.
 fn is_airflow_operator(segments: &[&str]) -> bool {
     is_airflow_builtin_or_provider(segments, "operators", "Operator")
 }
 
+/// Check whether the symbol is coming from the `log` builtin or provider module which ends
+/// with `TaskHandler`.
 fn is_airflow_task_handler(segments: &[&str]) -> bool {
     is_airflow_builtin_or_provider(segments, "log", "TaskHandler")
 }
```

Reply:
Additionally, as convention these helper functions would go at the bottom after the rule logic i.e., the rule definition would be up top and then the entrypoint to the rule logic and then all the helper structs and functions.

Reply:
@dhruvmanila  Thanks! This helps a ton! Just updated it and add more docstring to each utility functions